### PR TITLE
feat: trigger swap transactions

### LINKED ISSUE #20344: Notifications
We need to show proper toasts to keep the user updated on the approve and swap transactions

### LINKED ISSUE #20381: Trigger Swap
image

WE've got valid swap proposal, Swap button is enabled, we need to launch the actual Swap tx.

### PR DESCRIPTION
**Pull request by briansztamfater**
fixes #20381 
fixes #20344 

### Summary

This PR integrates swaps transactions actually triggering after user confirms the transaction. Also adds proper toasts notifications as per designs.

https://github.com/user-attachments/assets/ae5b0be9-575c-437c-b57e-b2c88a43cf13

#### Platforms

- Android
- iOS

##### Functional

- wallet / transactions

### Steps to test

- Open Status
- Login
- Go to wallet
- Select an account
- Select Swap option
- Select a ERC20 token on Select asset to pay screen
- Enter a valid amount
- Wait for swap proposal to appear
- Tap on "Review swap"
- Slide to execute the transction
- Enter password
- Swap modal should be dismissed and toast should be shown
- Wait for the swap transaction to be confirmed / failed and proper toasts are shown 

status: ready

### CODE REVIEW

**Comment by VolodLytvynenko** [comment]
hi @briansztamfater could you please rebase the current PR? thanx

**Comment by VolodLytvynenko** [comment]
@briansztamfater should the swap feature still be feature-flagged? In the latest build, I had to manually enable it in the feature flags to see the swap option

**Comment by VolodLytvynenko** [comment]
## PR_ISSUE 1: "Something went wrong, please try again later" Error message shown after entering valid value to swap

### Steps:
1. Go to swap page
2. Enter valid available value


### Actual result:
'Something went wrong, please try again later' error message is shown
image

### Expected result:
The user should be able to proceed to the swap confirmation page after entering a valid amount.


### OS: 
IOS, Android

### Devices:
- Pixel 7a, Android 13
- iPhone 11 Pro Max, IOS 17

### Logs
[logs (9).zip](https://github.com/user-attachments/files/16796284/logs.9.zip)

**Comment by VolodLytvynenko** [comment]
@briansztamfater I think some commits might be missing in the latest build because the swap UI looks completely different from what was shown in the [demo](https://github.com/status-im/status-mobile/pull/21134#issue-2490893221) compared with the latest build where the [issue](https://github.com/status-im/status-mobile/pull/21134#issuecomment-2317050581) is found. Could you please check?

File: src/status_im/contexts/wallet/signals.cljs
Comment (shivekkhurana) [inline src/status_im/contexts/wallet/signals.cljs:26]: Is `swap-transactions-id` a set ?

File: src/status_im/contexts/wallet/swap/events.cljs
Comment (shivekkhurana) [inline src/status_im/contexts/wallet/swap/events.cljs:350]: What is this sorcery 😅

File: src/status_im/contexts/wallet/signals.cljs
Comment (briansztamfater) [reply src/status_im/contexts/wallet/signals.cljs:26]: It is

**Comment by briansztamfater** [comment]
> @briansztamfater should the swap feature still be feature-flagged? In the latest build, I had to manually enable it in the feature flags to see the swap option

We should remove it at some point, I'll raise a new PR removing the flag 👍 

About issue 1, please try again, I have just updated the code. As we don't have a way (yet) to select the asset to receive, I hardcoded USDT for testing purposes, but SNT should be the default, which is a token available for all users. If you don't have USDT in your tokens list that may cause problems for now. Once #21140 is done that should be fine.

File: src/status_im/contexts/wallet/swap/events.cljs
Comment (briansztamfater) [reply src/status_im/contexts/wallet/swap/events.cljs:350]: Just checking if status of the transaction is the final one (either confirmed or failed). Saw this pattern in many places of the codebase, for example this: https://github.com/status-im/status-mobile/blob/35e71289dce758ba04e3772cf64bbf63bef59dd2/src/status_im/contexts/profile/settings/events.cljs#L28

**Comment by briansztamfater** [comment]
Also I will rebase, it will take some time as there is some PR chaining so I need to rebase the dependant PRs first

**Comment by VolodLytvynenko** [comment]
@briansztamfater thank you for clarification. Unfortunately I am blocked to test the swap flow fully due [status go issue ](https://github.com/status-im/status-go/pull/5784). 

Only some UI issues are found:

## ISSUE 2: [Android] Device's native keyboard appears on Swap page
### Steps:
Go to the swap page within the app on an Android device.

### Actual result:
The device's native keyboard automatically opens, resulting in two keyboards being displayed: the app's keyboard and the device's native keyboard.
image

### Expected result:
Only the app's keyboard should be displayed, without triggering the device's native keyboard.
image

### ENV:
Pixel 7a, Android 13

**Comment by VolodLytvynenko** [comment]
## ISSUE 3: Max value balances not updating according to selected account

### Steps:
1. Recover a user with at least 2 accounts.
2. Navigate to the swap page.
3. Select a different account on the swap page.

### Actual result:
The max value balances displayed do not update to reflect the newly selected account's balances

https://github.com/user-attachments/assets/59ddf14b-336d-4f14-8943-c33091d98388


### Expected result:
The max balances should update to reflect the selected account's available funds on the swap page.

**Comment by VolodLytvynenko** [comment]
@briansztamfater Not sure if [issue 2](https://github.com/status-im/status-mobile/pull/21134#issuecomment-2318412824) and [issue 3](https://github.com/status-im/status-mobile/pull/21134#issuecomment-2318489916) are within the scope of this PR. If not, I'll create them separately.

**Comment by briansztamfater** [comment]
@VolodLytvynenko thanks for testing! Issue 3 is out of scope of this PR. Can work on issue 2 at the meantime, until status-go issue is fixed and the whole flow could be tested.

**Comment by briansztamfater** [comment]
@VolodLytvynenko updated status-go version which includes changes in https://github.com/status-im/status-go/pull/5784

**Comment by briansztamfater** [comment]
@VolodLytvynenko issue 2 should be fixed now

**Comment by VolodLytvynenko** [comment]
hi @briansztamfater  Thank you for fixes! Unfortunately, I can't complete the transaction. I keep encountering an "insufficient balance" error toast every time, even though my account has enough tokens.

It seems we may be calculating the fee incorrectly. In MetaMask, the fee for an ETH -> SNT transaction is estimated at $17, while in Status, it's $3. My account has $5 (0.0021 ETH) for fees, which should be enough, but our fee calculation might be off.
image

Additionally, I can't confirm this transaction on other L2 networks, which would be cheaper, because the default SNT for receiving has low liquidity on L2 networks. As a result, there are no quotes available (the same behavior it both MetaMask and Status Desktop), leading to the "something went wrong" message on mobile.
image
 
I suggest either temporary changing the receiving token to one with higher liquidity, like DAI or USDC, which would reduce the transaction fee to around $0.05, or merging this PR as it is and rechecking the feature once we can manually change the receiving tokens. WDYT?

**Comment by briansztamfater** [comment]
@VolodLytvynenko thanks for testing and the valuable feedback. I've set DAI as the default token for now. I just tried locally and could swap USDT/DAI without any issues. Please let me know if swaps work this way, and we can try to figure out other issues in follow up PRs

**Comment by VolodLytvynenko** [comment]
Thank you for PR, no issues from my side, except this UI lower priority ones that can be considered as follow ups

### ISSUE 4: Incorrect placement of max value and incorrect button label on swap approval page
## Steps:
1. Go to the swap page.
2. Select an ERC20 token.
3. Enter a valid value.
4. Tap the "Approve" button.

### Actual Result:
The "Max Value" is located in the wrong place on the approval page.
The button label says "Slide to swap" instead of "Slide to sign."
image

### Expected Result:
The "Max Value" should be correctly placed according to the design.
The button label should correctly say "Slide to sign" on the approval page.
image

**Comment by briansztamfater** [comment]
Awesome @VolodLytvynenko, just fixed Issue 4 too :)
Waiting for an approval from the dev side now. `blocked` label can be removed, right?

**Review by shivekkhurana — APPROVED** [review]

**Comment by VolodLytvynenko** [comment]
## ISSUE 5: The input and receiving value not properly rounded
### Steps:
1. Go to the swap page and select ETH.
2. Enter a minimum value, such as 0.0000000001 ETH.

###Actual Result:
- The input value of 0.0000000001 ETH is less than $0.01 USD and is displayed without proper rounding.
- The receiving value is displayed in an incorrect format, such as 2.4712412341234e-7, without proper rounding.
image

### Expected Result:
Potential solutions:
- Crypto values should be rounded so that the final decimal place always represents a value between $0.01 and $0.10 USD, ensuring that small values are displayed appropriately.
OR
- The token value should be rounded to the "maximum number of decimal places" as specified in https://github.com/status-im/status-mobile/issues/18244

Examples:
- If the user enters a fiat value resulting in a token value with more than the allowed decimal places (e.g., 0.37 USD which equals 0.0001001... ETH), and taps the "show in token" button, the value in the input field should become 0.0001 ETH.)
- When ETH is priced at $365, four decimal places should be used for ETH because 0.0001 ETH = $0.036 USD, which is between $0.01 and $0.10 USD

**Comment by VolodLytvynenko** [comment]
@briansztamfater Should the swap transaction be included in the activity tab in the scope of this PR? Currently, swap transactions are not showing at all there.
### Actual result:
image

### Expected result:
https://www.figma.com/design/AD2JSKg0I8dZcylyiGa62O/Swap-for-Mobile?node-id=54-48281&node-type=INSTANCE&m=dev

**Comment by VolodLytvynenko** [comment]
> `blocked` label can be removed, right?

yes. just removed it. thanx

**Comment by briansztamfater** [comment]
> @briansztamfater Should the swap transaction be included in the activity tab in the scope of this PR? Currently, swap transactions are not showing at all there.
> 
> ### Actual result:
> image
> 
> ### Expected result:
> https://www.figma.com/design/AD2JSKg0I8dZcylyiGa62O/Swap-for-Mobile?node-id=54-48281&node-type=INSTANCE&m=dev

Definitely not included in the scope of this PR. We have an issue for it #20345, and we should also check with the team how the changes in the Activity implementation will affect how we display swaps transactions.

Will take a look at issue 5 👍

**Comment by briansztamfater** [comment]
@VolodLytvynenko  issue 5 should be fixed 👍

**Comment by VolodLytvynenko** [comment]
@briansztamfater You mentioned earlier that if a user doesn't have a receiving token (in our case DAI), some errors might occur. Is this the issue shown on record?

https://github.com/user-attachments/assets/f53993da-74ae-4402-a71d-eb9f9e193e97

### Logs:
[Status-debug-logs.zip](https://github.com/user-attachments/files/16851727/Status-debug-logs.zip)

**Comment by VolodLytvynenko** [comment]
> @briansztamfater You mentioned earlier that if a user doesn't have a receiving token (in our case DAI), some errors might occur. Is this the issue shown on record?
> 
>  dai.mp4 
> ### Logs:
> [Status-debug-logs.zip](https://github.com/user-attachments/files/16851727/Status-debug-logs.zip)

@briansztamfater Facing this error even when DAI is present in account. I will share this user in DM

## ISSUE 6: "undefined is not a function" error is shown after an attempt to swap ERC20 token
### Steps:
1. Select erc20 token during swap
2. Tap "approve" button

### Actual result:
image

### Additional info:
Perhaps the issue is related to the fee. I tested a scenario where the user doesn't have enough ETH to cover the fee or has too little ETH, which could be causing this problem.

**Comment by briansztamfater** [comment]
@VolodLytvynenko issue should be fixed now, seems I introduced the bug with the last fix 😅 
Let me know how testing goes now!

**Review by mohsen-ghafouri — APPROVED** [review]

**Comment by VolodLytvynenko** [comment]
[issue 5](https://github.com/status-im/status-mobile/pull/21134#issuecomment-2326680991) is not fixed fully. 
the value in the receiving section is still not rounded. However, this can be addressed in a follow-up

image

**Comment by VolodLytvynenko** [comment]
[issue 2](https://github.com/status-im/status-mobile/pull/21134#issuecomment-2318412824) is still reproducible. I missed one more scenario where it occurs: the keyboard appears when the app returns from the background.

### Steps:
1. Go to swap page
2. Go to background
3. return device background

### Actual result:
The device's native keyboard automatically opens, resulting in two keyboards being displayed: the app's keyboard and the device's native keyboard

https://github.com/user-attachments/assets/d88e41d1-fe13-4758-b855-5c7aa549bbbd


### Device:
Pixel 7a, Android 14

**Comment by VolodLytvynenko** [comment]
## ISSUE 7: "View" button available after and during ERC-20 Token approval, leading to errors

### Steps:
1. Go to swap page selecting ERC-20 token for swap
2. Enter the value and approve it 
3. Tap view button perfoem approving again

### Actual result:
The `view `button is shown that helps to perfrom approving one more time that leads to the error

https://github.com/user-attachments/assets/5a5073bb-d733-4bb4-99aa-97be50f37a30


### Expected result:
The `view` button is not shown
https://www.figma.com/design/AD2JSKg0I8dZcylyiGa62O/Swap-for-Mobile?node-id=61-21859&node-type=INSTANCE&m=dev

**Comment by VolodLytvynenko** [comment]
It looks like Paraswap struggles with certain decimal values on the Swap Page. To prevent this, I suggest implementing rounding for entered amounts on the swap page, similar to what was done on the send page in [PR #20915](https://github.com/status-im/status-mobile/pull/20915). This would round values so that the final decimal place represents a value between $0.01 and $0.10 USD, ensuring proper display of small amounts and avoiding issues with third-party integrations. WDYT, @xAlisher?

**Comment by briansztamfater** [comment]
@VolodLytvynenko issue 2 happening this way is weird, for now I am forcing keyboard dismiss on Android every time the input is focused. The trade-off of this approach is that the input will loose focus and cursor will not be shown. We can create an issue to do some further research on this. iOS behavior remains as usual.

Also regarding issue 7, I removed the View button as it is not shown in Swap designs.

**Comment by briansztamfater** [comment]
> It looks like Paraswap struggles with certain decimal values on the Swap Page. To prevent this, I suggest implementing rounding for entered amounts on the swap page, similar to what was done on the send page in [PR #20915](https://github.com/status-im/status-mobile/pull/20915). This would round values so that the final decimal place represents a value between $0.01 and $0.10 USD, ensuring proper display of small amounts and avoiding issues with third-party integrations. WDYT, @xAlisher?

IMO this deserves a bit more research, we can track it in a separate issue and add all the findings there, because I am not sure what are the actual errors (will add better error handling for mobile soon) and how is the best way to prevent them, if there is any. Also maybe this impacts Desktop implementation too.

**Comment by VolodLytvynenko** [comment]
> > It looks like Paraswap struggles with certain decimal values on the Swap Page. To prevent this, I suggest implementing rounding for entered amounts on the swap page, similar to what was done on the send page in [PR #20915](https://github.com/status-im/status-mobile/pull/20915). This would round values so that the final decimal place represents a value between $0.01 and $0.10 USD, ensuring proper display of small amounts and avoiding issues with third-party integrations. WDYT, @xAlisher?
> 
> IMO this deserves a bit more research, we can track it in a separate issue and add all the findings there, because I am not sure what are the actual errors (will add better error handling for mobile soon) and how is the best way to prevent them, if there is any. Also maybe this impacts Desktop implementation too.

@briansztamfater For sure. You've done a lot of work in this PR. This feature needs a separate follow-up.

**Comment by briansztamfater** [comment]
> [issue 5](https://github.com/status-im/status-mobile/pull/21134#issuecomment-2326680991) is not fixed fully. the value in the receiving section is still not rounded. However, this can be addressed in a follow-up
> 
> image

Yes, we can work on this on a follow up, and discuss the desired approach with designers. Not sure if we should round the value returned by the swap provider, but we can discuss it in a separate issue.

Let me know if there's something else to fix in this PR!

**Comment by VolodLytvynenko** [comment]
Hi @briansztamfater, thank you for your work. Issues are fixed. PR is ready to be merged

**Pull request merged**
